kernel/unwind.c

```diff
@@ -0,0 +1,582 @@
+static int get_def(const stack_frame *base, const cfa_row *row, int reg_or_cfa, uint64_t *out) {
+  dwarf_loc loc = (reg_or_cfa == -1) ? row->cfa : row->registers[reg_or_cfa];
+  switch (loc.rule) {
+    case DW_UNDEFINED:
+      return 0;
+      break;
+    case DW_SAME:
+      if (reg_or_cfa == -1) {
+        com_printf("BUG: Confusing register definition: CFA = SAME\n");
+        return -1;
+      }
+      if (!bit_array_get(base->regs_defined, reg_or_cfa)) {
+        com_printf("ERROR: Can't use rule SAME on undefined register %d\n", reg_or_cfa);
+        return -1;
+      }
+      *out = base->registers[reg_or_cfa];
+      break;
+    case DW_OFFSET:
+      /* ! LEBENSGEFAHR ! */
+      *out = *(uint64_t *)(base->cfa + loc.offset);
+      break;
+    case DW_REG:
+      *out = base->registers[loc.reg];
+      break;
+    case DW_CFA:
+      *out = base->registers[loc.reg] + loc.offset;
+      break;
+  }
+  return 1;
+}
+
+static int unwind_frame(stack_frame *base, const cfa_row *row, stack_frame *out) {
+  int ret = get_def(base, row, -1, &base->cfa);
+  if (ret < 0) { return ret; }
+  if (ret == 0) {
+    com_printf("ERROR: Need definition for CFA!\n");
+    return -1;
+  }
+
+  for (int i = 0; i < 17; ++i) {
+    int ret = get_def(base, row, i, &out->registers[i]);
+    if (ret < 0) {
+      return -1;
+    }
+    bit_array_set(out->regs_defined, i, ret);
+  }
+  out->registers[7] = base->cfa;
+  bit_array_set(out->regs_defined, 7, ret);
+  return 0;
+}
+
+/* Searches for the right FDE for a given program location and returns both that
+ * and the CIE for that FDE while we're at it. */
+int find_fde(uint64_t rip, eh_cie* cie_out, eh_fde* fde_out) {
+  head read_head;
+  read_head.data = _eh_frame_start;
+  read_head.remaining = _eh_frame_end - _eh_frame_start;
+
+  eh_cie current_cie;
+  memset(&current_cie, 0, sizeof(eh_cie));
+  while (read_head.remaining > 0) {
+    uint32_t short_length;
+    uint64_t long_length;
+    if (read_u32(&read_head, &short_length)) { return -1; }
+    if (short_length == (uint32_t)-1) {
+      if (read_u64(&read_head, &long_length)) { return -1; }
+    } else {
+      long_length = short_length;
+    }
+
+    if (read_head.remaining < long_length) { return -1; }
+
+    head subhead = recurse(&read_head, long_length);
+
+    /* tells us if we're about to read a CIE or an FDE */
+    uint32_t id;
+    if (read_u32(&subhead, &id)) { return -1; }
+    if (id == 0) {
+      if (read_cie(&subhead, &current_cie)) { return -1; }
+      continue;
+    }
+    eh_fde fde;
+    memset(&fde, 0, sizeof(eh_fde));
+    fde.cie = &current_cie;
+    if (read_fde(&subhead, &fde)) { return -1; }
+    if (rip >= fde.pc_begin && rip - fde.pc_begin < fde.pc_len) {
+      *cie_out = current_cie;
+      fde.cie = cie_out;
+      *fde_out = fde;
+      return 0;
+    }
+  }
+  return 1;
+}
+
+/* BIST: Let's see if we can just *parse* everything in eh_frame */
```

Comment:
If this is a test function it's perhaps best to have that explicitly in the name..?